Mathlib/LinearAlgebra/Basis/VectorSpace.lean

```diff
@@ -0,0 +1,281 @@
+
+/-- If `s` is a linear independent set of vectors, we can extend it to a basis. -/
+noncomputable def extend (hs : LinearIndependent K ((↑) : s → V)) :
+    Basis (hs.extend (subset_univ s)) K V :=
+  Basis.mk
+    (@LinearIndependent.restrict_of_comp_subtype _ _ _ id _ _ _ _ (hs.linearIndependent_extend _))
+    (SetLike.coe_subset_coe.mp <| by simpa using hs.subset_span_extend (subset_univ s))
+#align basis.extend Basis.extend
+
+theorem extend_apply_self (hs : LinearIndependent K ((↑) : s → V)) (x : hs.extend _) :
+    Basis.extend hs x = x :=
+  Basis.mk_apply _ _ _
+#align basis.extend_apply_self Basis.extend_apply_self
+
+@[simp]
+theorem coe_extend (hs : LinearIndependent K ((↑) : s → V)) : ⇑(Basis.extend hs) = ((↑) : _ → _) :=
+  funext (extend_apply_self hs)
+#align basis.coe_extend Basis.coe_extend
+
+theorem range_extend (hs : LinearIndependent K ((↑) : s → V)) :
+    range (Basis.extend hs) = hs.extend (subset_univ _) := by
+  rw [coe_extend, Subtype.range_coe_subtype, setOf_mem_eq]
+#align basis.range_extend Basis.range_extend
+
+-- Porting note: adding this to make the statement of `subExtend` more readable
+/-- Auxiliary definition: the index for the new basis vectors in `Basis.sumExtend`.
+
+The specific value of this definition should be considered an implementation detail.
+-/
+def sumExtendIndex (hs : LinearIndependent K v) : Set V :=
+  LinearIndependent.extend hs.to_subtype_range (subset_univ _) \ range v
+
+/-- If `v` is a linear independent family of vectors, extend it to a basis indexed by a sum type. -/
+noncomputable def sumExtend (hs : LinearIndependent K v) : Basis (ι ⊕ sumExtendIndex hs) K V :=
+  let s := Set.range v
+  let e : ι ≃ s := Equiv.ofInjective v hs.injective
+  let b := hs.to_subtype_range.extend (subset_univ (Set.range v))
+  (Basis.extend hs.to_subtype_range).reindex <|
+    Equiv.symm <|
+      calc
+        Sum ι (b \ s : Set V) ≃ Sum s (b \ s : Set V) := Equiv.sumCongr e (Equiv.refl _)
+        _ ≃ b :=
+          haveI := Classical.decPred (· ∈ s)
+          Equiv.Set.sumDiffSubset (hs.to_subtype_range.subset_extend _)
+#align basis.sum_extend Basis.sumExtend
+
+theorem subset_extend {s : Set V} (hs : LinearIndependent K ((↑) : s → V)) :
+    s ⊆ hs.extend (Set.subset_univ _) :=
+  hs.subset_extend _
+#align basis.subset_extend Basis.subset_extend
+
+section
+
+variable (K V)
+
+/-- A set used to index `Basis.ofVectorSpace`. -/
+noncomputable def ofVectorSpaceIndex : Set V :=
+  (linearIndependent_empty K V).extend (subset_univ _)
+#align basis.of_vector_space_index Basis.ofVectorSpaceIndex
+
+/-- Each vector space has a basis. -/
+noncomputable def ofVectorSpace : Basis (ofVectorSpaceIndex K V) K V :=
+  Basis.extend (linearIndependent_empty K V)
+#align basis.of_vector_space Basis.ofVectorSpace
+
+instance (priority := 100) _root_.Module.Free.of_divisionRing : Module.Free K V :=
+  Module.Free.of_basis (ofVectorSpace K V)
+#align module.free.of_division_ring Module.Free.of_divisionRing
```

Comment:
This lemma has migrated from `LinearAlgebra.FreeModule.Basic`, all the rest are from `LinearAlgebra.Basis`